hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -0,0 +1,323 @@
+   */
+  public ReferenceCounted<IOmMetadataReader> get(String key,
+      boolean skipActiveCheck) throws IOException {
+    // Atomic operation to initialize the OmSnapshot instance (once) if the key
+    // does not exist.
+    ReferenceCounted<IOmMetadataReader> rcOmSnapshot =
+        dbMap.computeIfAbsent(key, k -> {
+          LOG.info("Loading snapshot. Table key: {}", k);
+          try {
+            return new ReferenceCounted<>(cacheLoader.load(k), false, this);
+          } catch (OMException omEx) {
+            // Return null if the snapshot is no longer active
+            if (!omEx.getResult().equals(FILE_NOT_FOUND)) {
+              throw new IllegalStateException(omEx);
+            }
+          } catch (IOException ioEx) {
+            // Failed to load snapshot DB
+            throw new IllegalStateException(ioEx);
+          } catch (Exception ex) {
+            // Unexpected and unknown exception thrown from CacheLoader#load
+            throw new IllegalStateException(ex);
+          }
+          // Do not put the value in the map on exception
+          return null;
+        });
+
+    if (rcOmSnapshot == null) {
+      // The only exception that would fall through the loader logic above
+      // is OMException with FILE_NOT_FOUND.
+      throw new OMException("Snapshot table key '" + key + "' not found, "
+          + "or the snapshot is no longer active",
+          OMException.ResultCodes.FILE_NOT_FOUND);
+    }
+
+    // If the snapshot is already loaded in cache, the check inside the loader
+    // above is ignored. But we would still want to reject all get()s except
+    // when called from SDT (and some) if the snapshot is not active any more.
+    if (!omSnapshotManager.isSnapshotStatus(key, SNAPSHOT_ACTIVE) &&
+        !skipActiveCheck) {
+      throw new OMException("Unable to load snapshot. " +
+          "Snapshot with table key '" + key + "' is no longer active",
+          FILE_NOT_FOUND);
+    }
+
+    // Increment the reference count on the instance.
+    rcOmSnapshot.incrementRefCount();
+
+    // Remove instance from clean up list when it exists.
+    // TODO: [SNAPSHOT] Check thread safety with release()
+    pendingEvictionList.remove(rcOmSnapshot);
+
+    // Check if any entries can be cleaned up.
+    // At this point, cache size might temporarily exceed cacheSizeLimit
+    // even if there are entries that can be evicted, which is fine since it
+    // is a soft limit.
+    cleanup();
+
+    return rcOmSnapshot;
+  }
+
+  /**
+   * Release the reference count on the OmSnapshot instance.
+   * @param key snapshot table key
+   */
+  public void release(String key) {
+    ReferenceCounted<IOmMetadataReader> rcOmSnapshot = dbMap.get(key);
+    Preconditions.checkNotNull(rcOmSnapshot,
+        "Key '" + key + "' does not exist in cache");
+
+    if (rcOmSnapshot.decrementRefCount() == 0L) {
+      // Eligible to be closed, add it to the list.
+      pendingEvictionList.add(rcOmSnapshot);
```

Comment:
I think the refcount value check has to synchronized with the evictionList mods; otherwise some other thread can modify the refcount in between the two.

Reply:
done